{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Score multi-session events using the replay score from Davidson et al."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import warnings\n",
    "\n",
    "import nelpy as nel\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load experimental data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "datadirs = ['/Users/ckemere/Development/Data/Buzsaki']\n",
    "\n",
    "fileroot = next( (dir for dir in datadirs if os.path.isdir(dir)), None)\n",
    "# conda install pandas=0.19.2\n",
    "if fileroot is None:\n",
    "    raise FileNotFoundError('datadir not found')\n",
    "\n",
    "load_from_nel = True\n",
    "\n",
    "# load from nel file:\n",
    "if load_from_nel:\n",
    "    jar = nel.load_pkl(os.path.join(fileroot,'gor01vvp01_processed_speed.nel'))\n",
    "    exp_data = jar.exp_data\n",
    "    aux_data = jar.aux_data\n",
    "    del jar\n",
    "    \n",
    "    with pd.HDFStore(os.path.join(fileroot,'DibaMetadata.h5')) as store:\n",
    "        df = store.get('Session_Metadata')\n",
    "        df2 = store.get('Subset_Metadata')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define subset of sessions to score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating subset of 40 sessions\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>animal</th>\n",
       "      <th>month</th>\n",
       "      <th>day</th>\n",
       "      <th>time</th>\n",
       "      <th>track</th>\n",
       "      <th>segment</th>\n",
       "      <th>duration</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>n_placecells</th>\n",
       "      <th>n_PBEs</th>\n",
       "      <th>Notes</th>\n",
       "      <th>prescreen_z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>22-24-40</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>1620.0</td>\n",
       "      <td>203</td>\n",
       "      <td>61</td>\n",
       "      <td>301</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>22-24-40</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>1620.0</td>\n",
       "      <td>203</td>\n",
       "      <td>61</td>\n",
       "      <td>301</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>16-40-19</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>1330.0</td>\n",
       "      <td>117</td>\n",
       "      <td>46</td>\n",
       "      <td>277</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>16-40-19</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>1330.0</td>\n",
       "      <td>117</td>\n",
       "      <td>46</td>\n",
       "      <td>277</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>16-40-19</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>1180.0</td>\n",
       "      <td>117</td>\n",
       "      <td>43</td>\n",
       "      <td>150</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>16-40-19</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>1180.0</td>\n",
       "      <td>117</td>\n",
       "      <td>43</td>\n",
       "      <td>150</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>1-22-43</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>1012.0</td>\n",
       "      <td>203</td>\n",
       "      <td>62</td>\n",
       "      <td>117</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>1-22-43</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>1012.0</td>\n",
       "      <td>203</td>\n",
       "      <td>62</td>\n",
       "      <td>117</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>22-24-40</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>912.0</td>\n",
       "      <td>203</td>\n",
       "      <td>66</td>\n",
       "      <td>103</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>22-24-40</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>912.0</td>\n",
       "      <td>203</td>\n",
       "      <td>66</td>\n",
       "      <td>103</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>1-22-43</td>\n",
       "      <td>one</td>\n",
       "      <td>short</td>\n",
       "      <td>617.0</td>\n",
       "      <td>203</td>\n",
       "      <td>71</td>\n",
       "      <td>91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>1-22-43</td>\n",
       "      <td>one</td>\n",
       "      <td>short</td>\n",
       "      <td>617.0</td>\n",
       "      <td>203</td>\n",
       "      <td>71</td>\n",
       "      <td>91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>21-16-25</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>720.0</td>\n",
       "      <td>171</td>\n",
       "      <td>82</td>\n",
       "      <td>57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>21-16-25</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>720.0</td>\n",
       "      <td>171</td>\n",
       "      <td>82</td>\n",
       "      <td>57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>17-29-30</td>\n",
       "      <td>one</td>\n",
       "      <td>short</td>\n",
       "      <td>490.0</td>\n",
       "      <td>68</td>\n",
       "      <td>32</td>\n",
       "      <td>46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>17-29-30</td>\n",
       "      <td>one</td>\n",
       "      <td>short</td>\n",
       "      <td>490.0</td>\n",
       "      <td>68</td>\n",
       "      <td>32</td>\n",
       "      <td>46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>16-40-54</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>861.0</td>\n",
       "      <td>41</td>\n",
       "      <td>25</td>\n",
       "      <td>42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>16-40-54</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>861.0</td>\n",
       "      <td>41</td>\n",
       "      <td>25</td>\n",
       "      <td>42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>21-16-25</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>457.0</td>\n",
       "      <td>171</td>\n",
       "      <td>64</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>21-16-25</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>457.0</td>\n",
       "      <td>171</td>\n",
       "      <td>64</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>16-40-54</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>485.0</td>\n",
       "      <td>41</td>\n",
       "      <td>16</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>16-40-54</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>485.0</td>\n",
       "      <td>41</td>\n",
       "      <td>16</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>16-53-46</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>470.0</td>\n",
       "      <td>81</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>16-53-46</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>470.0</td>\n",
       "      <td>81</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>14-28-51</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>385.0</td>\n",
       "      <td>80</td>\n",
       "      <td>32</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>14-28-51</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>385.0</td>\n",
       "      <td>80</td>\n",
       "      <td>32</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>15-12-23</td>\n",
       "      <td>one</td>\n",
       "      <td>short</td>\n",
       "      <td>347.0</td>\n",
       "      <td>74</td>\n",
       "      <td>24</td>\n",
       "      <td>34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>15-12-23</td>\n",
       "      <td>one</td>\n",
       "      <td>short</td>\n",
       "      <td>347.0</td>\n",
       "      <td>74</td>\n",
       "      <td>24</td>\n",
       "      <td>34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>17-29-30</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>800.0</td>\n",
       "      <td>68</td>\n",
       "      <td>34</td>\n",
       "      <td>33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>17-29-30</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>800.0</td>\n",
       "      <td>68</td>\n",
       "      <td>34</td>\n",
       "      <td>33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>15-55-31</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>660.0</td>\n",
       "      <td>97</td>\n",
       "      <td>40</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>15-55-31</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>660.0</td>\n",
       "      <td>97</td>\n",
       "      <td>40</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>15-22-3</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>530.0</td>\n",
       "      <td>82</td>\n",
       "      <td>37</td>\n",
       "      <td>31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>15-22-3</td>\n",
       "      <td>two</td>\n",
       "      <td>long</td>\n",
       "      <td>530.0</td>\n",
       "      <td>82</td>\n",
       "      <td>37</td>\n",
       "      <td>31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>14-42-6</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>520.0</td>\n",
       "      <td>109</td>\n",
       "      <td>36</td>\n",
       "      <td>29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>14-42-6</td>\n",
       "      <td>one</td>\n",
       "      <td>long</td>\n",
       "      <td>520.0</td>\n",
       "      <td>109</td>\n",
       "      <td>36</td>\n",
       "      <td>29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>12-52-15</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>328.0</td>\n",
       "      <td>59</td>\n",
       "      <td>26</td>\n",
       "      <td>28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>vvp01</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>12-52-15</td>\n",
       "      <td>two</td>\n",
       "      <td>short</td>\n",
       "      <td>328.0</td>\n",
       "      <td>59</td>\n",
       "      <td>26</td>\n",
       "      <td>28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>15-55-31</td>\n",
       "      <td>one</td>\n",
       "      <td>short</td>\n",
       "      <td>410.0</td>\n",
       "      <td>97</td>\n",
       "      <td>44</td>\n",
       "      <td>27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>gor01</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>15-55-31</td>\n",
       "      <td>one</td>\n",
       "      <td>short</td>\n",
       "      <td>410.0</td>\n",
       "      <td>97</td>\n",
       "      <td>44</td>\n",
       "      <td>27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   animal  month  day      time track segment  duration  n_cells  \\\n",
       "58  gor01      6    9  22-24-40   two   short    1620.0      203   \n",
       "58  gor01      6    9  22-24-40   two   short    1620.0      203   \n",
       "35  gor01      6    7  16-40-19   two   short    1330.0      117   \n",
       "35  gor01      6    7  16-40-19   two   short    1330.0      117   \n",
       "36  gor01      6    7  16-40-19   two    long    1180.0      117   \n",
       "36  gor01      6    7  16-40-19   two    long    1180.0      117   \n",
       "73  gor01      6    9   1-22-43   one    long    1012.0      203   \n",
       "73  gor01      6    9   1-22-43   one    long    1012.0      203   \n",
       "59  gor01      6    9  22-24-40   two    long     912.0      203   \n",
       "59  gor01      6    9  22-24-40   two    long     912.0      203   \n",
       "72  gor01      6    9   1-22-43   one   short     617.0      203   \n",
       "72  gor01      6    9   1-22-43   one   short     617.0      203   \n",
       "6   gor01      6    8  21-16-25   two    long     720.0      171   \n",
       "6   gor01      6    8  21-16-25   two    long     720.0      171   \n",
       "0   vvp01      4    9  17-29-30   one   short     490.0       68   \n",
       "0   vvp01      4    9  17-29-30   one   short     490.0       68   \n",
       "83  vvp01      4    9  16-40-54   two    long     861.0       41   \n",
       "83  vvp01      4    9  16-40-54   two    long     861.0       41   \n",
       "5   gor01      6    8  21-16-25   two   short     457.0      171   \n",
       "5   gor01      6    8  21-16-25   two   short     457.0      171   \n",
       "82  vvp01      4    9  16-40-54   two   short     485.0       41   \n",
       "82  vvp01      4    9  16-40-54   two   short     485.0       41   \n",
       "14  gor01      6   12  16-53-46   two    long     470.0       81   \n",
       "14  gor01      6   12  16-53-46   two    long     470.0       81   \n",
       "28  vvp01      4   25  14-28-51   one    long     385.0       80   \n",
       "28  vvp01      4   25  14-28-51   one    long     385.0       80   \n",
       "51  vvp01      4   16  15-12-23   one   short     347.0       74   \n",
       "51  vvp01      4   16  15-12-23   one   short     347.0       74   \n",
       "1   vvp01      4    9  17-29-30   one    long     800.0       68   \n",
       "1   vvp01      4    9  17-29-30   one    long     800.0       68   \n",
       "61  gor01      6   12  15-55-31   one    long     660.0       97   \n",
       "61  gor01      6   12  15-55-31   one    long     660.0       97   \n",
       "75  gor01      6   13   15-22-3   two    long     530.0       82   \n",
       "75  gor01      6   13   15-22-3   two    long     530.0       82   \n",
       "50  gor01      6   13   14-42-6   one    long     520.0      109   \n",
       "50  gor01      6   13   14-42-6   one    long     520.0      109   \n",
       "41  vvp01      4   17  12-52-15   two   short     328.0       59   \n",
       "41  vvp01      4   17  12-52-15   two   short     328.0       59   \n",
       "60  gor01      6   12  15-55-31   one   short     410.0       97   \n",
       "60  gor01      6   12  15-55-31   one   short     410.0       97   \n",
       "\n",
       "    n_placecells  n_PBEs  Notes  prescreen_z  \n",
       "58            61     301    NaN          NaN  \n",
       "58            61     301    NaN          NaN  \n",
       "35            46     277    NaN          NaN  \n",
       "35            46     277    NaN          NaN  \n",
       "36            43     150    NaN          NaN  \n",
       "36            43     150    NaN          NaN  \n",
       "73            62     117    NaN          NaN  \n",
       "73            62     117    NaN          NaN  \n",
       "59            66     103    NaN          NaN  \n",
       "59            66     103    NaN          NaN  \n",
       "72            71      91    NaN          NaN  \n",
       "72            71      91    NaN          NaN  \n",
       "6             82      57    NaN          NaN  \n",
       "6             82      57    NaN          NaN  \n",
       "0             32      46    NaN          NaN  \n",
       "0             32      46    NaN          NaN  \n",
       "83            25      42    NaN          NaN  \n",
       "83            25      42    NaN          NaN  \n",
       "5             64      37    NaN          NaN  \n",
       "5             64      37    NaN          NaN  \n",
       "82            16      37    NaN          NaN  \n",
       "82            16      37    NaN          NaN  \n",
       "14            36      36    NaN          NaN  \n",
       "14            36      36    NaN          NaN  \n",
       "28            32      36    NaN          NaN  \n",
       "28            32      36    NaN          NaN  \n",
       "51            24      34    NaN          NaN  \n",
       "51            24      34    NaN          NaN  \n",
       "1             34      33    NaN          NaN  \n",
       "1             34      33    NaN          NaN  \n",
       "61            40      32    NaN          NaN  \n",
       "61            40      32    NaN          NaN  \n",
       "75            37      31    NaN          NaN  \n",
       "75            37      31    NaN          NaN  \n",
       "50            36      29    NaN          NaN  \n",
       "50            36      29    NaN          NaN  \n",
       "41            26      28    NaN          NaN  \n",
       "41            26      28    NaN          NaN  \n",
       "60            44      27    NaN          NaN  \n",
       "60            44      27    NaN          NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# restrict sessions to explore to a smaller subset\n",
    "min_n_placecells = 16\n",
    "min_n_PBEs = 27 # 27 total events ==> minimum 21 events in training set\n",
    "\n",
    "df2_subset = df2[(df2.n_PBEs >= min_n_PBEs) & (df2.n_placecells >= min_n_placecells)]\n",
    "\n",
    "sessions = df2_subset['time'].values.tolist()\n",
    "segments = df2_subset['segment'].values.tolist()\n",
    "\n",
    "print('Evaluating subset of {} sessions'.format(len(sessions)))\n",
    "\n",
    "df2_subset.sort_values(by=['n_PBEs', 'n_placecells'], ascending=[0,0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parallel scoring"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NOTE:** it is relatively easy (syntax-wise) to score each session as a parallel task, but since the Bayesian scoring takes such a long time to compute, we can be more efficient (higher % utilization) by further parallelizing over events, and not just over sessions. This further level of parallelization makes the bookkeeping a little ugly, so I provide the code for both approaches here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "n_jobs = 3 # set this equal to number of cores\n",
    "n_shuffles = 100 # 5000\n",
    "n_samples = 1000 # 35000\n",
    "w=3 # single sided bandwidth (0 means only include bin who's center is under line, 3 means a total of 7 bins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d54a510>, <tornado.concurrent.Future object at 0x1c3d545198>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-a2004f8e552c440d87701e5fe12bb56e']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c6c2fd268>, <tornado.concurrent.Future object at 0x1c5d4a8c18>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-ec31b4898c2a41f38822596bde2d685f']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d55eea0>, <tornado.concurrent.Future object at 0x1c3d556c18>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-3782d78bf6b74e56ae833f159b32d881']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d539400>, <tornado.concurrent.Future object at 0x1c1853ff60>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-5b6589c5a2cb41829dc9e554eaf8e5cb']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d4a3400>, <tornado.concurrent.Future object at 0x1c5d497400>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-8140460386904121873621fc77fbf2d1']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d4792f0>, <tornado.concurrent.Future object at 0x1c5d476160>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-ed4a946b524441a0ac9baf66f964d802']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d4a3048>, <tornado.concurrent.Future object at 0x1c5d485c88>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-50f1d88fdc1c4ea696d0a9ec2a5372e0']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d5677b8>, <tornado.concurrent.Future object at 0x1c3d566358>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-5d2ba1ac0bbb4d5cbfa0ed93d6f3368e']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d490950>, <tornado.concurrent.Future object at 0x1c5d485940>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-baa066cd6e4a41a9bbfbdc6255e43483']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c6c2f8268>, <tornado.concurrent.Future object at 0x1c5d497f28>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-055c65bcaf964de0a358ba375149a9cb']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d54ac80>, <tornado.concurrent.Future object at 0x1c3d5454a8>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-b3c4a2a48c1f4c15983d0c8f48f905cc']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d54a158>, <tornado.concurrent.Future object at 0x1c3d52fe48>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-b03a577635ce40459b6446ce7f8b2f69']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c6c2f8a60>, <tornado.concurrent.Future object at 0x1c5d4a8898>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-275531550b144a108e58f0e534f8da62']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c6c2f8e18>, <tornado.concurrent.Future object at 0x1c5d4a8a58>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-2df943db086543de989e06b75263777d']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c6c2fd620>, <tornado.concurrent.Future object at 0x1c5d4a8dd8>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-2115127b8c44423b8d95905721238f0b']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d4901e0>, <tornado.concurrent.Future object at 0x1c5d485080>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-8d78b69b01734b25a29b3cd263bdfdeb']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d567b70>, <tornado.concurrent.Future object at 0x1c3d5666a0>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-1a2e9db9d84a46508b14b5db61af0d23']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d567f28>, <tornado.concurrent.Future object at 0x1c3d566860>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-5800d88f380f415f80b8d4f23274fe2e']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d471ea0>, <tornado.concurrent.Future object at 0x1c3d566f60>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-40ddc44529c942b3b1499c03be6c2974']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d554268>, <tornado.concurrent.Future object at 0x1c3d566550>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-e8e67141bb614bd1950a1ff70e577c7f']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d4831e0>, <tornado.concurrent.Future object at 0x1c5d476b00>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-db499f0fa8ca44fb9053ae3664884685']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d5397b8>, <tornado.concurrent.Future object at 0x1c3d52fa90>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-729fbd6c45a3428ea9115f1d05e02941']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d471378>, <tornado.concurrent.Future object at 0x1c3d566a20>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-36561eccba2d492abaeeb555a679e7b0']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d54a8c8>, <tornado.concurrent.Future object at 0x1c3d5452e8>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-d54bf2ab341d4ec39799df23638e3fd3']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d496510>, <tornado.concurrent.Future object at 0x1c5d4a86d8>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-f0db28ea908642f88f0691523fb1cc91']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c6c2f8620>, <tornado.concurrent.Future object at 0x1c5d4a8320>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-fc0a729e12d24ae190e4bd54d5029845']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d471ae8>, <tornado.concurrent.Future object at 0x1c3d566da0>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-efb72801150d4c749974620474b90d9c']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d483598>, <tornado.concurrent.Future object at 0x1c5d476d30>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-2dd3b31202404de58e7d40acb61dd76e']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d490598>, <tornado.concurrent.Future object at 0x1c5d4852e8>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-bbfdb79a90504cc199b65f1c35eebab5']\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c3d55eae8>, <tornado.concurrent.Future object at 0x1c3d545748>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-d967ac4c8d1547898b8eedfb3120ece5']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tornado.application - ERROR - Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x1c5d471730>, <tornado.concurrent.Future object at 0x1c3d566be0>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 605, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 277, in null_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/ioloop.py\", line 626, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1063, in run\n",
      "    yielded = self.gen.throw(*exc_info)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/joblib.py\", line 128, in callback_wrapper\n",
      "    result = yield _wait([future])\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1055, in run\n",
      "    value = future.result()\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/concurrent.py\", line 238, in result\n",
      "    raise_exc_info(self._exc_info)\n",
      "  File \"<string>\", line 4, in raise_exc_info\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/tornado/gen.py\", line 1069, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/ckemere/anaconda3/lib/python3.6/site-packages/distributed/client.py\", line 2906, in _wait\n",
      "    raise CancelledError(cancelled)\n",
      "concurrent.futures._base.CancelledError: ['work_events-batch-cd04fce818024cfc97a13f364b1c748c']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=3)]: Done   1 tasks      | elapsed:    0.2s\n"
     ]
    },
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'joblib'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-6952777e3058>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     27\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mparallel_backend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'dask.distributed'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mscheduler_host\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'104.154.255.65:8786'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[0;31m# Anything returned by work() can be stored:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 29\u001b[0;31m     \u001b[0mparallel_results\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mParallel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m51\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdelayed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwork_events\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparallel_events\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/joblib/parallel.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m    787\u001b[0m                 \u001b[0;31m# consumption.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    788\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_iterating\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 789\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mretrieve\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    790\u001b[0m             \u001b[0;31m# Make sure that we get a last message telling us we are done\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    791\u001b[0m             \u001b[0melapsed_time\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_start_time\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/joblib/parallel.py\u001b[0m in \u001b[0;36mretrieve\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    699\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_output\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    700\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 701\u001b[0;31m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_output\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    702\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    703\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mBaseException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexception\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mresult\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    164\u001b[0m                       self._result, raiseit=False, callback_timeout=timeout)\n\u001b[1;32m    165\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'error'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 166\u001b[0;31m             \u001b[0msix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreraise\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    167\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'cancelled'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/six.py\u001b[0m in \u001b[0;36mreraise\u001b[0;34m(tp, value, tb)\u001b[0m\n\u001b[1;32m    683\u001b[0m             \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    684\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__traceback__\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mtb\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 685\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    686\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    687\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/work/miniconda/lib/python3.6/site-packages/distributed/protocol/pickle.py\u001b[0m in \u001b[0;36mloads\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'joblib'"
     ]
    }
   ],
   "source": [
    "# Parallelize by EVENT\n",
    "import dask\n",
    "import distributed.joblib\n",
    "from joblib import Parallel, delayed \n",
    "from joblib import parallel_backend\n",
    "\n",
    "# A function that can be called to do work:\n",
    "def work_events(arg):    \n",
    "\n",
    "    # Split the list to individual variables:\n",
    "    session, segment, ii, bst, tc = arg\n",
    "    scores, shuffled_scores, percentiles = nel.analysis.replay.score_Davidson_final_bst_fast(bst=bst,\n",
    "                                                                                        tuningcurve=tc,\n",
    "                                                                                        w=w,\n",
    "                                                                                        n_shuffles=n_shuffles,\n",
    "                                                                                        n_samples=n_samples)\n",
    "\n",
    "    return (session, segment, ii, scores, shuffled_scores, percentiles)\n",
    "\n",
    "# List of instances to pass to work():\n",
    "# unroll all events:\n",
    "parallel_events = []\n",
    "for session, segment in zip(sessions, segments):\n",
    "    for nn in range(aux_data[session][segment]['PBEs'].n_epochs):\n",
    "        parallel_events.append((session, segment, nn, aux_data[session][segment]['PBEs'][nn], aux_data[session][segment]['tc']))\n",
    "\n",
    "with parallel_backend('dask.distributed', scheduler_host='104.154.255.65:8786'):\n",
    "    # Anything returned by work() can be stored:\n",
    "    parallel_results = Parallel(n_jobs=n_jobs, verbose=51)(map(delayed(work_events), parallel_events))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "# standardize parallel results\n",
    "bdries_ = [aux_data[session][segment]['PBEs'].n_epochs for session, segment in zip(sessions, segments) ]\n",
    "bdries = np.cumsum(np.insert(bdries_,0,0))\n",
    "bdries\n",
    "\n",
    "sessions_ = np.array([result[0] for result in parallel_results])\n",
    "segments_ = np.array([result[1] for result in parallel_results])\n",
    "idx = [result[2] for result in parallel_results]\n",
    "\n",
    "scores_bayes_evt = np.array([float(result[3]) for result in parallel_results])\n",
    "scores_bayes_shuffled_evt = np.array([result[4].squeeze() for result in parallel_results])\n",
    "scores_bayes_percentile_evt = np.array([float(result[5]) for result in parallel_results])\n",
    "\n",
    "results = {}\n",
    "for nn in range(len(bdries)-1):\n",
    "    session = np.unique(sessions_[bdries[nn]:bdries[nn+1]])\n",
    "    if len(session) > 1:\n",
    "        raise ValueError(\"parallel results in different format / order than expected!\")\n",
    "    session = session[0]\n",
    "    segment = np.unique(segments_[bdries[nn]:bdries[nn+1]])\n",
    "    if len(segment) > 1:\n",
    "        raise ValueError(\"parallel results in different format / order than expected!\")\n",
    "    segment = segment[0]\n",
    "    try:\n",
    "        results[session][segment]['scores_bayes'] = scores_bayes_evt[bdries[nn]:bdries[nn+1]]\n",
    "    except KeyError:\n",
    "        try:\n",
    "            results[session][segment] = dict()\n",
    "            results[session][segment]['scores_bayes'] = scores_bayes_evt[bdries[nn]:bdries[nn+1]]\n",
    "        except KeyError:\n",
    "            results[session] = dict()\n",
    "            results[session][segment] = dict()\n",
    "            results[session][segment]['scores_bayes'] = scores_bayes_evt[bdries[nn]:bdries[nn+1]]\n",
    "\n",
    "    results[session][segment]['scores_bayes_shuffled'] = scores_bayes_shuffled_evt[bdries[nn]:bdries[nn+1]]\n",
    "    results[session][segment]['scores_bayes_percentile'] = scores_bayes_percentile_evt[bdries[nn]:bdries[nn+1]]\n",
    "\n",
    "print('done packing results')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save results to disk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "jar = nel.ResultsContainer(results=results, description='gor01 and vvp01 speed restricted results for best 20 candidate sessions')\n",
    "jar.save_pkl('score_bayes_all_sessions.nel')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
